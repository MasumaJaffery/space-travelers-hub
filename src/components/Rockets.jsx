import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getRockets, cancelReservation, reserveRocket } from '../redux/RocketSlice';
import 'bootstrap/dist/css/bootstrap.min.css';

const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.rockets);

  useEffect(() => {
    if (rockets.length === 0) {
      dispatch(getRockets());
    }
  }, [dispatch, rockets]);

  return (
    <div>
      <h1>Rockets</h1>
      <div className="card mb-3" style={{ maxWidth: '90vw' }}>
        <div className="rockets__container d-flex-column">
          {rockets.length === 0 ? (
            <h2>Loading info...</h2>
          ) : (
            <ul className="rocketslist">
              {rockets.map((rocket) => (
                <li key={rocket.id} className="rocketsitems">
                  <div className="row g-0">
                    <div className="col-md-4">
                      <img
                        src={rocket.flickr_images[0]}
                        alt={rocket.name}
                        className="img-fluid rounded-start"
                      />
                    </div>
                    <div className="col-md-8">
                      <div className="card-body">
                        <h3 className="card-title">{rocket.rocket_name}</h3>
                        <p className="card-text">
                          {rocket.description}
                        </p>
                        <div>
                          <button
                            type="button"
                            className={rocket.reserved ? 'rockets__cancel-btn' : 'rockets__reserve-btn btn btn-primary'}
                            onClick={() => {
                              dispatch(rocket.reserved
                                ? cancelReservation(rocket.id)
                                : reserveRocket(rocket.id));
                            }}
                          >
                            {rocket.reserved ? 'Cancel Reservation' : 'Reserve Rocket'}
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </li>
              ))}
            </ul>
          )}
        </div>
      </div>
    </div>
  );
};

export default Rockets;
