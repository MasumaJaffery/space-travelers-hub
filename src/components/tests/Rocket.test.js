import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Rockets from '../Rockets';

test('renders loading message when rockets list is empty', () => {
  render(<Rockets />);
  const loadingMessage = screen.getByText(/Loading info.../i);
  expect(loadingMessage).toBeInTheDocument();
});

test('renders rockets when rockets list is not empty', () => {
  const mockRockets = [
    {
      id: 'rocket1',
      rocket_name: 'Falcon 1',
      description: 'Description 1',
      reserved: false,
      flickr_images: ['image1.jpg'],
    },
  ];

  render(<Rockets rockets={mockRockets} />);
  const rocketName = screen.getByText(/Falcon 1/i);
  expect(rocketName).toBeInTheDocument();
});

test('clicking reserve button toggles reservation status', () => {
  const mockRocket = {
    name: 'Falcon 1',
    description: 'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.',
    id: '5e9d0d95eda69955f709d1eb',
    reserved: false,
    flickr_images: [
      'https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg',
      'https://farm4.staticflickr.com/3955/32915197674_eee74d81bb_b.jpg',
      'https://farm1.staticflickr.com/293/32312415025_6841e30bf1_b.jpg',
      'https://farm1.staticflickr.com/623/23660653516_5b6cb301d1_b.jpg',
      'https://farm6.staticflickr.com/5518/31579784413_d853331601_b.jpg',
      'https://farm1.staticflickr.com/745/32394687645_a9c54a34ef_b.jpg',
    ],
  };

  render(<Rockets rockets={[mockRocket]} />);
  const reserveButton = screen.getByText(/Reserve Rocket/i);
  fireEvent.click(reserveButton);

  const cancelReservationButton = screen.getByText(/Cancel Reservation/i);
  expect(cancelReservationButton).toBeInTheDocument();
});
