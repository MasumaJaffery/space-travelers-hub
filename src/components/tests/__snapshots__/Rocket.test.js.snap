// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Empty Missions component correctly 1`] = `
<div>
  <div>
    <div
      class="card mb-3"
      style="max-width: 90vw;"
    >
      <div
        class="rockets__container d-flex-column"
      >
        <ul
          class="rocketslist"
        >
          <li
            class="rocketsitems"
          >
            <div
              class="row g-0"
            >
              <div
                class="col-md-4"
              >
                <img
                  alt="Falcon 1"
                  class="img-fluid rounded-start"
                  src="https://imgur.com/DaCfMsj.jpg"
                />
              </div>
              <div
                class="col-md-8"
              >
                <div
                  class="card-body"
                >
                  <h3
                    class="card-title"
                  />
                  <p
                    class="card-text"
                  >
                    <div
                      class="btn-rocket"
                    />
                    The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
                  </p>
                  <div>
                    <button
                      class="rockets__reserve-btn btn btn-primary"
                      type="button"
                    >
                      Reserve Rocket
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </li>
          <li
            class="rocketsitems"
          >
            <div
              class="row g-0"
            >
              <div
                class="col-md-4"
              >
                <img
                  alt="Falcon 9"
                  class="img-fluid rounded-start"
                  src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
                />
              </div>
              <div
                class="col-md-8"
              >
                <div
                  class="card-body"
                >
                  <h3
                    class="card-title"
                  />
                  <p
                    class="card-text"
                  >
                    <div
                      class="btn-rocket"
                    />
                    Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
                  </p>
                  <div>
                    <button
                      class="rockets__reserve-btn btn btn-primary"
                      type="button"
                    >
                      Reserve Rocket
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
